dex/src/main/java/io/crate/data/BatchIterator.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data;
+
+import java.util.concurrent.CompletableFuture;
+
+
+/**
+ * A batch iterator used to navigate over batched results. Users of this interface must not rely on any thread safety of
+ * the underlying implementation.
+ */
+public interface BatchIterator extends Row {
```

Comment:
pls implement an AssertingBatchIterator which could be used to wrap in tests and use it in tests. this proxy can then be used to describe the assertions that are in this interface

however i think the interface documentations should cover exactly what happens on moveNext when false is returned (e.g: the position tracking usecase) 

also what happens to the row methods when move* methods are called and return false